docs/design/2021-03-30-small-table-caching.md

```diff
@@ -0,0 +1,174 @@
+
+If the update operation is not supported at all, the use case would be too limited. So even if we trade writing ability for reading performance, we should at least support updating.
+
+### Key algorithm
+
+There is a conceptual "read-write" lock to implement the caching operation.
+
+Before caching the data, the table needs to hold a "read lock", so that the data will not be modified afterward. This "read lock" is a record providing some meta-information. All modifications must check the lock meta-information first. If the record is in the read-locked state, modifications are forbidden.
+
+The "read" lock needs to maintain a lease for a while, and the lease should be renewed continuously. The underlying data will be safe as long as the read lock lease hold.
+
+When performing a modification, the meta-information should be checked first. If there is a "read" lock, the "read" lock needs to be upgraded to a "write intend" lock. This step prevents the read lock from renewing the lease. After the lease expires, the meta status is change to "write" lock, write operation can be performed afterward. And after the write operation finishes, the "write" lock should be cleared, so that the subsequent "read" lock can be added and the data can be cached again.
+
+The "write" lock also needs a TTL(time to live) to handle abnormal situations. For example, if the TiDB server crashes after the write lock succeeds, TTL gives us a hint of how to clean the orphan write lock.
+
+Put all things together, we have this interface:
+
+```
+type StateRemote interface {
+	Load()
+    LockForRead()
+    LockForWrite()
+    RenewLease()
+    WriteAndUnlock()
+}
+```
+
+### Management of lock meta information
+
+What’s the granularity of the lock? One option is to implement the meta lock at the key-value level. The advantage is that the granularity is finer and the updating performance is better. If we cache the key-value record, modifying one key would not affect the cache of the other keys.  The disadvantage is that this will invade the current transaction layer implementation, and make our transaction more complex. Considering the current implementation of the transaction is very complex already, it is not wise to do so.
+
+The other option is to implement the meta lock at the table level. Each table corresponds to one meta lock record. As long as any key of the table is modified, the table cache becomes invalid. Assuming that we only deal with small tables, and this table is updated infrequently, this option should be acceptable.
+
+For updating operation, the `StateRemote` interface needs to be implemented. The following part of this section describes the management of lock meta information, where it should be stored, and how it is used.
+
+Such a table could be used to store the meta-information:
+
+```
+CREATE TABLE mysql.table_cache_meta (
+   tid int primary key,
+   lock_type enum('NONE','READ','INTEND','WRITE'),
+   lease uint64,
+   oldReadLease uint64,
```

Comment:
May be `old_read_lease ` is better to have same format with `lock_type` . But we should also update the exist code.

Reply:
Yes ... until recently I find this silly mistake: the naming style ...
But update this code would add another bootstrap version which is bad